File: vagtplan.ts
import { createEvents, EventAttributes } from 'ics'
import * as XLSX from 'sheetjs'

const months = [
  'januar',
  'februar',
  'marts',
  'april',
  'maj',
  'juni',
  'juli',
  'august',
  'september',
  'oktober',
  'november',
  'december',
]

const ignoredTurns = /^(DATO|FRI|FERIE|BAGBAG)$/i
const initials = /(^|\s)SG(\s|$)/i

export function main(path: string): EventAttributes[] {
  const ws = XLSX.readFile(path)
  const rows = XLSX.utils.sheet_to_csv(ws.Sheets['Ark1']).split('\n')
  const columns = rows[1].split(',')
  const dateColumn = columns.indexOf('DATO')

  const m = rows[0].match(/(\w+) (\d+)/)
  if (!m) throw new Error('Could not parse month')
  const month = months.indexOf(m[1].toLowerCase()) + 1
  const year = parseInt(m[2])

  const events: EventAttributes[] = []
  for (const row of rows.slice(2)) {
    const values = row.split(',')
    const day = parseInt(values[dateColumn])
    if (isNaN(day)) break
    let hours = 0

    let start = Temporal.ZonedDateTime.from({
      timeZone: 'Europe/Copenhagen',
      year,
      month,
      day,
      hour: 8,
      minute: 15,
    }).withTimeZone('UTC')
    if ([6, 7].includes(start.dayOfWeek)) {
      start = start.add(Temporal.Duration.from({ minutes: 45 }))
    }

    for (let i = dateColumn; i < columns.length; i++) {
      if (columns[i].match(ignoredTurns)) continue
      if (!values[i].match(initials)) continue
      let minutes = 0

      switch (columns[i]) {
        case 'BA.VA':
          switch (start.dayOfWeek) {
            case 5:
              hours = 24
              minutes = 45
              break
            case 7:
              hours = 23
              minutes = 15
              break
            default:
              hours = 24
          }
          break
        case 'STUEGANG':
          if (hours > 8) continue
        default:
          hours = 8
      }

      events.push({
        start: [start.year, start.month, start.day, start.hour, start.minute],
        startInputType: 'utc',
        duration: { hours, minutes },
        title: columns[i],
      })
    }
  }

  return events
}

if (import.meta.main) {
  const extension = /.xlsx?$/
  if (!Deno.args[0]?.match(extension)) {
    console.error('Usage: vagtplan.exe <excel file>')
    Deno.exit(1)
  }
  Deno.writeTextFileSync(
    Deno.args[0].replace(extension, '.ics'),
    createEvents(main(Deno.args[0])).value ?? '',
  )
}

File: vagtplan.go
package main

import (
	"log"
	"os"
	"regexp"
	"slices"
	"strconv"
	"strings"
	"time"

	ics "github.com/arran4/golang-ical"
	"github.com/google/uuid"
	"github.com/xuri/excelize/v2"
)

var months = []string{
	"januar",
	"februar",
	"marts",
	"april",
	"maj",
	"juni",
	"juli",
	"august",
	"september",
	"oktober",
	"november",
	"december",
}

var ignoredTurns = regexp.MustCompile(`^(DATO|FRI|FERIE|BAGBAG)$`)
var initials = regexp.MustCompile(`(^|\s)SG(\s|$)`)

func main() {
	if len(os.Args) != 2 {
		log.Fatal("Usage: vagtplan.exe <xlsx file>")
	}

	ws, _ := excelize.OpenFile(os.Args[1])
	rows, _ := ws.GetRows("Ark1")
	columns := rows[1]
	dateColumn := slices.Index(columns, "DATO")

	m := regexp.MustCompile(`(\w+) (\d+)`).FindStringSubmatch(rows[0][5])
	if m == nil {
		log.Fatal("Could not parse month")
	}
	month := slices.Index(months, strings.ToLower(m[1])) + 1
	year, _ := strconv.Atoi(m[2])

	cal := ics.NewCalendar()
	loc, _ := time.LoadLocation("Europe/Copenhagen")
	for _, row := range rows[2:] {
		day, err := strconv.Atoi(row[dateColumn])
		if err != nil {
			break
		}
		hours := 0

		start := time.Date(year, time.Month(month), day, 8, 15, 0, 0, loc).UTC()
		if start.Weekday() == time.Saturday || start.Weekday() == time.Sunday {
			start = start.Add(time.Minute * 45)
		}

		for i, cell := range row {
			if ignoredTurns.MatchString(columns[i]) {
				continue
			}
			if !initials.MatchString(cell) {
				continue
			}
			minutes := 0

			switch columns[i] {
			case "BA.VA":
				switch start.Weekday() {
				case time.Friday:
					hours = 24
					minutes = 45
				case time.Sunday:
					hours = 23
					minutes = 15
				default:
					hours = 24
				}
			case "STUEGANG":
				if hours > 8 {
					continue
				}
				hours = 8
			default:
				hours = 8
			}

			event := cal.AddEvent(uuid.NewString())
			event.SetSummary(columns[i])
			event.SetStartAt(start)
			event.SetEndAt(start.Add(time.Duration(hours)*time.Hour + time.Duration(minutes)*time.Minute))
		}
	}

	icsFile := strings.Replace(os.Args[1], ".xlsx", ".ics", 1)
	file, _ := os.Create(icsFile)
	defer file.Close()

	file.Write([]byte(cal.Serialize()))
}
